### DISCUSSION TITLE
RAG Fusion

### DISCUSSION THREAD

**Discussion by mikehudson2** [FlowiseAI/Flowise#1807] [Ideas]
Is it possible to create a "RAG Fusion" workflow in Flowise?  i.e. take a query, use the LLM to generate multiple similar queries, submit all those queries to the vector database, re-rank/de-dupe the returned chunks and then send what remains to the LLM to generate an answer? As per [https://github.com/Raudaschl/rag-fusion](url)  and [https://arxiv.org/pdf/2402.03367.pdf](url)

**Comment by Jaredude** [comment]
You can accomplish this fairly easily using multiple chains and a flow that uses Custom Tool with Conversational Retrieval Agent. For example, you'd create a flow that performs a keyword/phrase expansion and then chain the output of that to another LLM chain that condenses the phrases by removing similar phrases. Then call that chain from a custom tool that takes that input and passes that into your database or API. The output of the tool is then used by the Conversational Retrieval Agent to craft the RAG fusion result.

**Comment by mikehudson2** [comment]
Thanks, I'll have a look at that. In the meantime I got something working using the "VectorStore to Document" document loader.
![21   RAG Fusion](https://github.com/FlowiseAI/Flowise/assets/38255405/6497b92a-3ce7-4975-8b0b-41e052a1702c)
